### PR TITLE
Changes to DatastreamVerification

### PR DESCRIPTION
**Pull request by awead**
This solves two problems: 1) XML datastreams that were being altered on migration are now checked using a checksum calculated from their Nokogiri documents; and 2) datastreams that had no checksum in Fedora3 are now compared with a newly calculated one.

### CODE REVIEW

**Comment by awead** [comment]
@cam156 FYI. and @mjgiarlo if you want to take a look.

File: lib/fedora_migrate/datastream_verification.rb [hunk 1]
```diff
@@ -4,28 +4,20 @@ module FedoraMigrate::DatastreamVerification
...
     @datastream = datastream || @source
-    check = has_no_checksum? || has_matching_checksums? || content_equivalent?
+    check = has_matching_checksums? || has_matching_nokogiri_checksums?
     FedoraMigrate::Logger.warn "#{@datastream.pid} datastream #{@datastream.dsid} validation failed" unless check
     check
   end
 
   def has_matching_checksums?
-    datastream.checksum == target_checksum
+    datastream.checksum == target_checksum || Digest::SHA1.hexdigest(datastream.content) == target_checksum
```
Comment (mjgiarlo) [inline lib/fedora_migrate/datastream_verification.rb]: Under what condition is the latter disjunct evaluated?  As in, when would `datastream.checksum == target_checksum` return false or nil?

File: lib/fedora_migrate/datastream_verification.rb [hunk 2]
```diff
@@ -4,28 +4,20 @@ module FedoraMigrate::DatastreamVerification
...
-    end
-  end
-  
-  # TODO: In some cases, the line <?xml version="1.0"?> is being added to migrated xml datastreams.
-  # This invalidates the checksum even though the xml content is the same. (issue #11)
-  def content_equivalent?
+  def has_matching_nokogiri_checksums?
     if datastream.mimeType == "text/xml"
-      EquivalentXml.equivalent?(target_content, datastream.content)
+      Digest::SHA1.hexdigest(Nokogiri::XML(datastream.content).to_xml) == Digest::SHA1.hexdigest(Nokogiri::XML(target_content).to_xml)
```
Comment (mjgiarlo) [inline lib/fedora_migrate/datastream_verification.rb]: How about creating a method named something like `checksum` in this file or elsewhere if there's a better place in the fedora-migrate codebase? That way you don't have to type "Digest::SHA1.hexdigest" over and over again, and it also makes it easier to folks to override this value with their own checksumming method.

File: lib/fedora_migrate/datastream_verification.rb [hunk 3]
```diff
@@ -4,28 +4,20 @@ module FedoraMigrate::DatastreamVerification
...
+    else
+      false
```
Comment (mjgiarlo) [inline lib/fedora_migrate/datastream_verification.rb]: You can also rewrite this as a guard clause and save LOC and branchiness:

```ruby
def has_matching_nokogiri_checksums?
  return false unless datastream.mimeType == 'text/xml'
  checksum(Nokogiri::XML(datastream.content).to_xml) == checksum(Nokogiri::XML(target_content))
end

File: spec/integration/object_migration_spec.rb [hunk 4]
```diff
@@ -27,6 +27,18 @@
       expect(subject.edit_users).to include("jilluser@example.com")
     end
 
+    describe "objects with Om datastreams" do
+      let(:mover) { FedoraMigrate::ObjectMover.new source, ExampleModel::OmDatastreamExample.new }
```
Comment (mjgiarlo) [inline spec/integration/object_migration_spec.rb]: This is subjective, but I'd prefer to see explicit parens for the outer `#new` call here. It confused my eyes for a minute.

File: spec/unit/datastream_verification_spec.rb [hunk 5]
```diff
@@ -10,40 +10,61 @@ def initialize datastream
...
-    subject do
-      expect_any_instance_of(TestSubject).to receive(:target_checksum).once.and_return("foo")
-      TestSubject.new(mock_source)
+  describe "binary sources from Fedora3" do
+    let(:bad_binary_source)     { double("Datastream", checksum: "bad",     mimeType: "binary", content: "XXXXXX", dsid: "content", pid: "abc123") }
+    let(:good_binary_source)    { double("Datastream", checksum: "foo",     mimeType: "binary", content: "foo",    dsid: "content", pid: "abc123") }
+    let(:missing_checksum)      { double("Datastream", checksum: "missing", mimeType: "binary", content: "foo",    dsid: "content", pid: "abc123") }
+    context "that match Fedora4's checksum" do
+      subject do
+        expect_any_instance_of(TestSubject).to receive(:target_checksum).once.and_return("foo")
```
Comment (mjgiarlo) [inline spec/unit/datastream_verification_spec.rb]: Do you want a test here or a stub?

File: spec/unit/datastream_verification_spec.rb [hunk 6]
```diff
@@ -10,40 +10,61 @@ def initialize datastream
...
-  end
-
-  context "when the checksum is 'none'" do
-    let(:mock_source) { double("Datastream", checksum: "none") }
-    subject do
-      TestSubject.new(mock_source)
+    context "that do not match Fedora4's checksum" do
+      subject do
+        expect_any_instance_of(TestSubject).to receive(:target_checksum).twice.and_return("bar")
+        expect(FedoraMigrate::Logger).to receive(:warn)
```
Comment (mjgiarlo) [inline spec/unit/datastream_verification_spec.rb]: Same as above re: expect statements in subject blocks.

File: spec/unit/datastream_verification_spec.rb [hunk 7]
```diff
@@ -10,40 +10,61 @@ def initialize datastream
...
-  context "with equivalent content" do
-    let(:mock_source) { double("Datastream", checksum: "bad", mimeType: "text/xml", content: "<bar></bar>") }
-    subject do
-      expect_any_instance_of(TestSubject).to receive(:target_content).once.and_return("<?xml version=\"1.0\"?>\n<bar></bar>")
-      TestSubject.new(mock_source)
+    context "when the checksum is missing" do
+      subject { TestSubject.new(missing_checksum) }
+      context "and a newly calculated checksum matches" do
+        before do
+          expect_any_instance_of(TestSubject).to receive(:target_checksum).twice.and_return(Digest::SHA1.hexdigest("foo"))
```
Comment (mjgiarlo) [inline spec/unit/datastream_verification_spec.rb]: Same as above re: expect statements in subject blocks.

File: spec/unit/datastream_verification_spec.rb [hunk 8]
```diff
@@ -10,40 +10,61 @@ def initialize datastream
...
+        end
+        it { is_expected.to have_matching_checksums }
+        it { is_expected.to be_valid }
+      end
+      context "and a newly calculated checksum does not match" do
+        before do
+          expect_any_instance_of(TestSubject).to receive(:target_checksum).twice.and_return(Digest::SHA1.hexdigest("bar"))
```
Comment (mjgiarlo) [inline spec/unit/datastream_verification_spec.rb]: Same as above re: expect statements in before blocks.

File: spec/unit/datastream_verification_spec.rb [hunk 9]
```diff
@@ -10,40 +10,61 @@ def initialize datastream
...
-    let(:mock_source) { double("Datastream", checksum: "bad", mimeType: "binary", content: "XXXXXX", dsid: "content", pid: "abc123") }
-    subject do
-      expect_any_instance_of(TestSubject).to receive(:target_checksum).once.and_return("foo")
-      expect(FedoraMigrate::Logger).to receive(:warn).once
-      TestSubject.new(mock_source)
+  describe "xml sources from Fedora3" do
+    let(:xml_source) { double("Datastream", checksum: "invalid", mimeType: "text/xml", content: "<bar></bar>") }
+    context "when the datastream content is correctly altered upon migration" do
+      subject do
+        expect_any_instance_of(TestSubject).to receive(:target_content).once.and_return("<?xml version=\"1.0\"?>\n<bar></bar>")
```
Comment (mjgiarlo) [inline spec/unit/datastream_verification_spec.rb]: Same as above re: expect statements in subject blocks.

File: spec/unit/datastream_verification_spec.rb [hunk 10]
```diff
@@ -10,40 +10,61 @@ def initialize datastream
...
+        TestSubject.new(xml_source)
+      end
+      it { is_expected.to have_matching_nokogiri_checksums }
+    end
+    context "when the datastream content is incorrectly altered upon migration" do
+      subject do
+        expect_any_instance_of(TestSubject).to receive(:target_content).once.and_return("<?xml version=\"1.0\"?>\n<baz></baz>")
```
Comment (mjgiarlo) [inline spec/unit/datastream_verification_spec.rb]: Same as above re: expect statements in subject blocks.

File: lib/fedora_migrate/datastream_verification.rb (hunk 1 above)
Comment (awead) [inline lib/fedora_migrate/datastream_verification.rb]: When the checksum from Fedora3 is "NONE" or if the checksum actually did fail. So we calculate a new one based on the original content from Fedora3.

**Comment by awead** [comment]
@mjgiarlo yo :dog:

**Comment by mjgiarlo** [comment]
:+1:

File: lib/fedora_migrate/datastream_verification.rb [hunk 11]
```diff
@@ -4,29 +4,18 @@ module FedoraMigrate::DatastreamVerification
...
     @datastream = datastream || @source
-    check = has_no_checksum? || has_matching_checksums? || content_equivalent?
+    check = has_matching_checksums? || has_matching_nokogiri_checksums?
     FedoraMigrate::Logger.warn "#{@datastream.pid} datastream #{@datastream.dsid} validation failed" unless check
     check
   end
 
   def has_matching_checksums?
-    datastream.checksum == target_checksum
+    datastream.checksum == target_checksum || checksum(datastream.content) == target_checksum
```
Comment (mjgiarlo) [inline lib/fedora_migrate/datastream_verification.rb]: Like I said, :+1:, though I might like to see this method look like:

```ruby
def has_matching_checksums?
  checksum(datastream) == target_checksum
end
```

```ruby
def checksum content
  content.respond_to?(:checksum) ? content.checksum : Digest::SHA1.hexdigest(content)
end
```

But it's :peach:-y as is!

**Pull request merged**